### ISSUE TITLE
How to check if there's a translation

### ISSUE THREAD

**Issue by nWidart** [dimsav/laravel-translatable#108]
Hello,

This might be a stupid question, but I'd like to check if there's a translation before echoing, to avoid the `Trying to get property of non-object` error.

I've tried:

``` php
isset($page->translate($lang)->title) ? $page->translate($lang)->title : ''
is_object($page->translate($lang)) ? $page->translate($lang)->title : ''
$page->translate($lang) ? $page->translate($lang)->title : ''
```
None work.

Since the a certain version, when a translation doesn't exists, it doesn't return an empty string. It might be nice to return an empty string of no translation exists in the db for that language, to avoid having to make these checks in views all the time.

Thank you.

**Comment by sdebacker** [comment]
what about ``` $model->hasTranslation($locale) ``` ?

**Comment by nWidart** [comment]
Oh! Thank works indeed. :+1: 

I still think this would be nice to have built in, if no translation exist return empty string.
Since this logic needs to be repeated -every- time.

Thanks! :smile:

**Comment by dimsav** [comment]
@nWidart what is your use case?

`Null` should be fine too. At least for the views.

**Comment by nWidart** [comment]
Well for instance on this [view](https://github.com/AsgardCms/Page/blob/develop/Resources/views/admin/partials/edit-fields.blade.php#L16) I have to constantly check for the presence of the translation before actually using it, to avoid the `Trying to get property of non object` error.

It would be nice if I hadn't had to use the `hasTranslation` check, but could directly use `$page->translate($lang)->slug` without having to worry about the presence or not of a translation.

**Comment by dimsav** [comment]
Do you really need custom $language? If not, you could do this: `$page->slug`.

**Comment by nWidart** [comment]
I do, since I display all data separated in tabs for different languages.

![screen shot 2015-03-26 at 3 48 57 pm](https://cloud.githubusercontent.com/assets/882397/6848804/b745b4ba-d3cf-11e4-97f0-f2c9602fb41f.png)

Otherwise, on the front-end, indeed `$page->slug` is sufficient.

**Comment by dimsav** [comment]
Makes sense. We could create a friendly helper method for that. Any suggestions?

Brainstorming: `$page->t('slug', 'default')`

**Comment by nWidart** [comment]
That would be nice indeed. What do you think @sdebacker ?

Out of curiosity, why not make the `->translate()` method return empty string if no translation available ?

**Comment by dimsav** [comment]
Because you would have `$page->translate()->name` where `$page->translate()` is a string. (fatal error).

**Comment by nWidart** [comment]
Ah I see :smiley:

**Comment by sdebacker** [comment]
The old behavior also miss me. Nice to have a helper !

**Comment by dimsav** [comment]
Damn. `$page->t('slug', 'default')` wouldn't work (locale is missing).

What about `$page->t('fr', 'name')` ? I still don't like the syntax :/

**Comment by dimsav** [comment]
Or maybe `$page->t('name', 'fr');` where second parameter defaults to current locale.

**Comment by nWidart** [comment]
`$page->t('fr', 'name', 'default')` then ?

Where the 3rd parameter would optional, and defaults to `''` ?

**Comment by nWidart** [comment]
Hm if locale is second, that would be the `default` value would need an empty second param, or make the default value second param.

`$page->t('name', 'fr', 'default');`

**Comment by dimsav** [comment]
What do you need the default for? I think this is cleaner than a third parameter.

```
{{ $page->t('name', 'fr') ?: 'default' }}
```

or even this:

```
{{ $page->t('name', 'fr') or 'default' }}
```

**Comment by nWidart** [comment]
Ah okay, I was just suggesting since you added a default parameter in an earlier response.

That would just make more logic in the view, and having a default parameter moves that logic in your package. It would just be nice if it'd return `''` (empty string) if no translation is available.

**Comment by sdebacker** [comment]
What about returning null when there is no record for that language in DB, and return empty string when there is an empty record.

**Comment by nWidart** [comment]
@sdebacker That would work too, I think. Echoing an `null` value would just be empty I guess. :smiley:

**Comment by dimsav** [comment]
Sounds good. Anyone volunteering for this? I created a new branch 'next-major' for version 6.

**Comment by nWidart** [comment]
You mean to develop it ? I'm quite busy at the moment. 
I'd gladly test it though when available. :smiley:

**Comment by llioor** [comment]
Any news about $page->t('name', 'fr') ? (I think it is the best option)

**Comment by dimsav** [comment]
@nWidart @llioor it is possible to use this syntax to avoid the "Trying to get property of non-object" error:

```php
$page->translateOrNew('fr')->name;
```

translateOrNew('fr') will return an empty translation instance if a french translation doesn't exist. And translateOrNew('fr')->name will return `null`.

I would like to avoid introducing a method like `t()` for no serious reason.

**Comment by nWidart** [comment]
Thanks for the update. I wasn't a big fan of the `t()` method anyway, it's not very expressive / showing intent.

I would have preferred it working 'out of the box', with the current api. Ie if `$page->translate('en')->name` doesn't have a translation in the current language, return null. This would avoid having to change every single view on the new api. 

But I guess it's ok. If asking for an api change, what's the advantage of using `translateOrNew()` over `hasTranslation()` ? All I'm seeing is one less condition I guess.

**Comment by joshbrw** [comment]
Just my +1, I always use ->translateOrNew()

**Comment by dimsav** [comment]
We could consider implementing that change but in a new major update only.

**Comment by sdebacker** [comment]
For me it should return null when there is no record in translation table.
Like in query builder, ```first()``` and ```find($id)``` both return null when there are no items.

**Comment by sdebacker** [comment]
I'm happy with the package like it works for the moment. I previously said that the old behavior miss me, but it's not the case anymore. About nWidart’s question about avoiding to repeat $page->hasTranslation($locale) in forms, I solved it with this package: https://github.com/TypiCMS/TranslatableBootForms
